File: 4-median-of-two-sorted-arrays/median-of-two-sorted-arrays.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        A,B = nums1, nums2

        if len(A) > len(B):
            A,B = B,A

        total = len(A)+len(B)

        half = total//2

        l,r  = 0, len(A) - 1

        while True:
            midA = (l+r)//2
            midB = half - midA-2

            Aleft = A[midA] if midA >= 0 else float('-inf')
            Aright = A[midA+1] if (midA+1) < len(A) else float('inf')

            Bleft = B[midB] if midB >= 0 else float('-inf')
            Bright = B[midB+1] if (midB+1) < len(B) else float('inf')

            if Aleft <= Bright and Bleft <= Aright:
                if total % 2 ==0:
                    return (max(Aleft,Bleft) + min(Bright,Aright))/2

                return min(Aright,Bright)

            elif Aleft> Bright:
                r = midA-1 

            else:
                l  = midA + 1

File: 4-median-of-two-sorted-arrays/median-of-two-sorted-arrays.java
class Solution 
{
    public double findMedianSortedArrays(int[] nums1, int[] nums2) 
    {
        // O(log(m + n))
    
        
        // let nums1 be the shorter one
        if (nums2.length < nums1.length) return findMedianSortedArrays(nums2, nums1);
        
        // define the number of elements from nums1 that are put into the final left side
        int lo = 0;
        int hi = nums1.length;
        int N = nums1.length + nums2.length;
        
        // if N is
        // odd:  count(left) == count(right) + 1
        // even: count(left) == count(right)
        // let assume, the final state is there is p1 elements from nums1
        // and p2 elements from nums2 and there is:
        // 1: p1 + p2 == (N + 1) / 2
        // 2: nums1[p1 - 1] <= nums2[p2]
        // 3: nums2[p2 - 1] <= nums1[p1]
        // and the median is:
        // odd:  Math.max(nums1[p1 - 1], nums2[p2 - 1]) if possible
        // even: (Math.max(nums1[p1 - 1], nums2[p2 - 1]) + Math.min(nums1[p1], nums2[p2])) / 2.0 if possible
        // in both above cases, if any index is invalid then just don't use it ..
        while (lo < hi)
        {
            int p1 = (lo + hi) / 2;
            int p2 = (N + 1) / 2 - p1;
        
            if (p1 == 0 || nums1[p1 - 1] <= nums2[p2])
            {
                if (p2 == 0 || nums2[p2 - 1] <= nums1[p1]) break;
                else lo = p1 + 1;
            }
            else hi = p1 - 1;
        }
        
        int p1 = (lo + hi) / 2;
        int p2 = (N + 1) / 2 - p1;
        int m = Integer.MIN_VALUE;
        int n = Integer.MAX_VALUE;
        if (N % 2 == 1)
        {
            return Math.max(p1 - 1 >= 0 ? nums1[p1 - 1] : m, p2 - 1 >= 0 ? nums2[p2 - 1] : m);
        }
        else
        {
            return (Math.max(p1 - 1 >= 0 ? nums1[p1 - 1] : m, p2 - 1 >= 0 ? nums2[p2 - 1] : m) + 
                    Math.min(p1 < nums1.length ? nums1[p1] : n, p2 < nums2.length ? nums2[p2] : n)) / 2.0;
        }
    }
}